src/preprocessing/passes/bv_to_int.cpp

```diff
@@ -787,19 +786,118 @@ Node BVToInt::bvToInt(Node n)
   return d_bvToIntCache[n];
 }
 
-bool BVToInt::childrenTypesChanged(Node n) {
+void BVToInt::defineBVUFAsIntUF(Node bvUF)
+{
+  // This function should only be called after translating
+  // the function symbol to a new function symbol
+  // with the right domain and range.
+  Assert(d_bvToIntCache.find(bvUF) != d_bvToIntCache.end());
+
+  // get domain and range of the original function
+  TypeNode tn = bvUF.getType();
+  vector<TypeNode> bvDomain = tn.getArgTypes();
+  TypeNode bvRange = tn.getRangeType();
+
+  // get the translated function symbol
+  Node intUF = d_bvToIntCache[bvUF];
+
+  // create a symbolic  application to be used in define-fun
+
+  // symbolic arguments of original function
+  vector<Expr> args;
+  // children of the new symbolic application
+  vector<Node> achildren;
+  achildren.push_back(intUF);
+  int i = 0;
+  for (TypeNode d : bvDomain)
+  {
+    // Each bit-vector argument is casted to a natural number
+    // Other arguments are left intact.
+    Node fresh_bound_var = d_nm->mkBoundVar(d);
+    args.push_back(fresh_bound_var.toExpr());
+    Node casted_arg = castIfNeeded(args[i], d_nm->integerType());
+    achildren.push_back(casted_arg);
+    i++;
+  }
+  Node intApplication = d_nm->mkNode(kind::APPLY_UF, achildren);
+  // If the range is BV, the application needs to be casted back.
+  intApplication = castIfNeeded(intApplication, bvRange);
+  // add the function definition to the smt engine.
+  smt::currentSmtEngine()->defineFunction(
+      bvUF.toExpr(), args, intApplication.toExpr(), true);
+}
+
+bool BVToInt::childrenTypesChanged(Node n)
+{
   bool result = false;
-  for (Node child : n) {
+  for (Node child : n)
+  {
     TypeNode originalType = child.getType();
     TypeNode newType = d_bvToIntCache[child].get().getType();
-    if (! newType.isSubtypeOf(originalType)) {
+    if (!newType.isSubtypeOf(originalType))
+    {
       result = true;
       break;
     }
   }
   return result;
 }
 
+Node BVToInt::castIfNeeded(Node n, TypeNode tn)
```

Comment:
I think it'd be clearer if this method was merged with castToType (e.g. castToType always did these checks).